pkg/validator/app.go

```diff
@@ -27,24 +27,63 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// validateUniqueness checks the lists of roles and service IDs for duplicates.
+// Any generated error messages will be added to the input list and returned.
+func validateUniqueness(
+	allRoleIDs []string,
+	allServiceIDs []string,
+	valErrors []string,
+) []string {
+
+	if !shared.ListIsUnique(allRoleIDs) {
+		valErrors = append(valErrors, nonUniqueRoleID)
+	}
+	if !shared.ListIsUnique(allServiceIDs) {
+		valErrors = append(valErrors, nonUniqueServiceID)
+	}
+	return valErrors
+}
```

Comment:
why not mutate by removing the duplicates

Reply:
Just because the ID is duplicate doesn't mean that the whole service or role object is duplicate. If e.g. there are two different role objects with the ID "controller", we don't know which one should be correct, so I think we just have to reject the whole thing.